demo/src/faber.ts

```diff
@@ -0,0 +1,168 @@
+    return await this.agent.connections.receiveInvitation(invitationMessage)
+  }
+
+  private async waitForConnection(connectionRecord: ConnectionRecord) {
+    connectionRecord = await this.agent.connections.returnWhenIsConnected(connectionRecord.id)
+    console.log(`${Color.green}${Output.connectionEstablished}${Color.reset}`)
+    return connectionRecord.id
+  }
+
+  public async acceptConnection(invitation_url: string) {
+    const connectionRecord = await this.receiveConnectionRequest(invitation_url)
+    if (connectionRecord === undefined) {
+      return
+    }
+    this.connectionRecordAliceId = await this.waitForConnection(connectionRecord)
+  }
+
+  private printSchema(name: string, version: string, attributes: string[]) {
+    console.log(`\n\nThe credential definition will look like this:\n`)
+    console.log(`${Color.purlpe}Name: ${Color.reset}${name}`)
+    console.log(`${Color.purlpe}Version: ${Color.reset}${version}`)
+    console.log(`${Color.purlpe}Attributes: ${Color.reset}${attributes[0]}, ${attributes[1]}, ${attributes[2]}\n`)
+  }
+
+  private async registerSchema() {
+    const schemaTemplate = {
+      name: 'Faber College' + uuid(),
+      version: '1.0.0',
+      attributes: ['name', 'degree', 'date'],
+    }
+    this.printSchema(schemaTemplate.name, schemaTemplate.version, schemaTemplate.attributes)
+    this.ui.updateBottomBar(`${Color.green}\nRegistering schema...\n`)
+    const schema = await this.agent.ledger.registerSchema(schemaTemplate)
+    this.ui.updateBottomBar('\nSchema registerd!\n')
+    return schema
+  }
+
+  private async registerCredentialDefiniton(schema: Schema) {
+    this.ui.updateBottomBar('\nRegistering credential defenition...\n')
+    this.credentialDefinition = await this.agent.ledger.registerCredentialDefinition({
+      schema,
+      tag: 'latest',
+      supportRevocation: false,
+    })
+    this.ui.updateBottomBar('\nCredential definition registerd!!\n')
+    return this.credentialDefinition
+  }
+
+  private getCredentialPreview() {
+    const credentialPreview = CredentialPreview.fromRecord({
+      name: 'Alice Smith',
+      degree: 'Computer Science',
+      date: '01/01/2022',
+    })
+    return credentialPreview
+  }
+
+  public async issueCredential() {
+    const schema = await this.registerSchema()
+    const credentialDefinition = await this.registerCredentialDefiniton(schema)
+    const credentialPreview = this.getCredentialPreview()
+    const connectionRecord = await this.getConnectionRecord()
+
+    this.ui.updateBottomBar('\nSending credential offer...\n')
+    await this.agent.credentials.offerCredential(connectionRecord.id, {
+      credentialDefinitionId: credentialDefinition.id,
+      preview: credentialPreview,
+    })
+    this.ui.updateBottomBar(`\nCredential offer sent! \n${Color.reset}`)
+  }
+
+  private async printProofFlow(print: string) {
+    this.ui.updateBottomBar(print)
+    await new Promise((f) => setTimeout(f, 2000))
```

Comment:
Why do we need to wait here?

Reply:
Because if you don't, the program is too fast for printing to see what is going on.